### PR TITLE
[WIP] Sync python interpreters

### PR DESCRIPTION
**Pull request by bissenbay**
Related-To: https://github.com/thoth-station/storages/issues/933

### CODE REVIEW

**Comment by goern** [comment]
recheck

**Comment by bissenbay** [comment]
> recheck

Locally, all tests are passing.

File: thoth/storages/data/alembic/versions/b23e6145a6ab_added_a_model_haspythoninterpreter_to_.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    op.create_table('has_python_interpreter',
+    sa.Column('id', sa.Integer(), autoincrement=True, nullable=False),
+    sa.Column('path', sa.String(length=256), nullable=False),
+    sa.Column('link', sa.String(length=256), nullable=True),
+    sa.Column('version', sa.String(length=256), nullable=True),
+    sa.Column('package_extract_run_id', sa.Integer(), nullable=False),
+    sa.Column('software_environment_id', sa.Integer(), nullable=False),
+    sa.ForeignKeyConstraint(['package_extract_run_id'], ['package_extract_run.id'], ondelete='CASCADE'),
+    sa.ForeignKeyConstraint(['software_environment_id'], ['software_environment.id'], ondelete='CASCADE'),
+    sa.PrimaryKeyConstraint('id', 'package_extract_run_id', 'software_environment_id')
```
Comment (fridex) [inline thoth/storages/data/alembic/versions/b23e6145a6ab_added_a_model_haspythoninterpreter_to_.py]: Not sure if this is the right approach - you are creating relation on 3 tables.

What about creating a table which would carry python interpreter information (path, version, link) and link to it from `package_extract_run` table? `has_python_interpreter` would be than capturing many-to-many relationship.

File: thoth/storages/graph/models.py [hunk 2]
```diff
@@ -269,6 +270,20 @@ class FoundPythonFile(Base, BaseExtension):
...
+
+    __tablename__ = "has_python_interpreter"
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+
+    path = Column(String(256), nullable=False)
+    link = Column(String(256), nullable=True)
+    version = Column(String(256), nullable=True)
+
+    package_extract_run = relationship("PackageExtractRun", back_populates="has_python_interpreter")
```
Comment (bissenbay) [inline thoth/storages/graph/models.py]: Creating a relationship to `PackageExtractRun`.

File: thoth/storages/graph/models.py [hunk 3]
```diff
@@ -247,6 +247,7 @@ class PackageExtractRun(Base, BaseExtension):
     python_package_version_entities = relationship("Identified", back_populates="package_extract_run")
     software_environment = relationship("SoftwareEnvironment", back_populates="package_extract_runs")
     versioned_symbols = relationship("DetectedSymbol", back_populates="package_extract_run")
+    has_python_interpreter = relationship("HasPythonInterpreter", back_populates="package_extract_run")
```
Comment (bissenbay) [inline thoth/storages/graph/models.py]: Creating a relationship to `HasPythonInterpreter`.

File: thoth/storages/graph/postgres.py [hunk 4]
```diff
@@ -3112,6 +3113,21 @@ def _python_file_digests_sync_analysis_result(self, package_extract_run: Package
                 file=py_file["filepath"],
             )
 
+    def _python_interpreters_sync_analysis_result(
```
Comment (bissenbay) [inline thoth/storages/graph/postgres.py]: Syncing the data.

File: thoth/storages/graph/postgres.py [hunk 5]
```diff
@@ -3112,6 +3114,27 @@ def _python_file_digests_sync_analysis_result(self, package_extract_run: Package
...
+        self,
+        package_extract_run: PackageExtractRun,
+        document: dict,
+        software_environment: SoftwareEnvironment or ExternalSoftwareEnvironment
```
Comment (fridex) [inline thoth/storages/graph/postgres.py]:
```python
        software_environment: Union[SoftwareEnvironment, ExternalSoftwareEnvironment]
```

File: thoth/storages/graph/models.py [hunk 6]
```diff
@@ -242,6 +242,7 @@ class PackageExtractRun(Base, BaseExtension):
     )
 
     found_python_files = relationship("FoundPythonFile", back_populates="package_extract_run")
+    found_python_interpreter = relationship("FoundPythonInterpreter", back_populates="package_extract_run")
```
Comment (fridex) [inline thoth/storages/graph/models.py]: I've just spotted this - we try to respect plural in many-to-many relations, so please adjust to `found_python_interpreters` (here and bellow).

File: thoth/storages/graph/models.py [hunk 7]
```diff
@@ -242,6 +242,7 @@ class PackageExtractRun(Base, BaseExtension):
     )
 
     found_python_files = relationship("FoundPythonFile", back_populates="package_extract_run")
+    found_python_interpreters = relationship("FoundPythonInterpreter", back_populates="package_extract_run")
```
Comment (bissenbay) [inline thoth/storages/graph/models.py]: Changed to `found_python_interpreters`.

File: thoth/storages/graph/models.py [hunk 8]
```diff
@@ -269,6 +270,34 @@ class FoundPythonFile(Base, BaseExtension):
...
+
+    __tablename__ = "found_python_interpreter"
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+
+    python_interpreter_id = Column(Integer, ForeignKey("python_interpreter.id", ondelete="CASCADE"), primary_key=True)
+    package_extract_run_id = Column(Integer, ForeignKey("package_extract_run.id", ondelete="CASCADE"), primary_key=True)
+
+    python_interpreter = relationship("PythonInterpreter", back_populates="package_extract_runs")
+    package_extract_run = relationship("PackageExtractRun", back_populates="found_python_interpreters")
```
Comment (bissenbay) [inline thoth/storages/graph/models.py]: Changed to `found_python_interpreters`.

File: thoth/storages/graph/postgres.py [hunk 9]
```diff
@@ -3112,6 +3114,27 @@ def _python_file_digests_sync_analysis_result(self, package_extract_run: Package
                 file=py_file["filepath"],
             )
 
+    def _python_interpreters_sync_analysis_result(
+        self,
+        package_extract_run: PackageExtractRun,
+        document: dict,
+        software_environment: Union[SoftwareEnvironment, ExternalSoftwareEnvironment]
```
Comment (fridex) [inline thoth/storages/graph/postgres.py]: This parameter is unused now.

Comment (bissenbay) [reply thoth/storages/graph/postgres.py]: @fridex , do you want me to delete it?

Comment (fridex) [reply thoth/storages/graph/postgres.py]: I wonder why linters do not report anything here. Let's keep it then, we will need it in future :) :+1:

**Pull request merged**